lib/systemd.c

```diff
@@ -20,68 +20,50 @@
  */
 
 #include <zebra.h>
+#include <sys/un.h>
 
 #include "thread.h"
 #include "systemd.h"
 
-#if defined HAVE_SYSTEMD
-#include <systemd/sd-daemon.h>
-#endif
-
-/*
- * Wrapper this silliness if we
- * don't have systemd
+/* these are cleared from env so they don't "leak" into things we fork(),
+ * particularly for watchfrr starting individual daemons
+ *
+ * watchdog_pid is currently not used since watchfrr starts forking.
+ * (TODO: handle that better, somehow?)
  */
+static pid_t watchdog_pid = -1;
+static intmax_t watchdog_usec;
+
+/* not used yet, but can trigger auto-switch to journald logging */
+bool sd_stdout_is_journal;
+bool sd_stderr_is_journal;
+
+static char *notify_socket;
+
+/* talk to whatever entity claims to be systemd ;) */
 static void systemd_send_information(const char *info)
 {
-#if defined HAVE_SYSTEMD
-	sd_notify(0, info);
-#else
-	return;
-#endif
-}
+	int sock;
+	struct sockaddr_un sun;
 
-/*
- * A return of 0 means that we are not watchdoged
- */
-static int systemd_get_watchdog_time(int the_process)
-{
-#if defined HAVE_SYSTEMD
-	uint64_t usec;
-	char *watchdog = NULL;
-	int ret;
-
-	ret = sd_watchdog_enabled(0, &usec);
-
-	/*
-	 * If return is 0 -> we don't want watchdog
-	 * if return is < 0, some sort of failure occurred
-	 */
-	if (ret < 0)
-		return 0;
-
-	/*
-	 * systemd can return that this process
-	 * is not the expected sender of the watchdog timer
-	 * If we set the_process = 0 then we expect to
-	 * be able to send the watchdog to systemd
-	 * irrelevant of the pid of this process.
-	 */
-	if (ret == 0 && the_process)
-		return 0;
-
-	if (ret == 0 && !the_process) {
-		watchdog = getenv("WATCHDOG_USEC");
-		if (!watchdog)
-			return 0;
-
-		usec = atol(watchdog);
-	}
+	if (!notify_socket)
+		return;
+
+	sock = socket(AF_UNIX, SOCK_DGRAM, 0);
+	if (sock < 0)
+		return;
+
+	sun.sun_family = AF_UNIX;
+	strlcpy(sun.sun_path, notify_socket, sizeof(sun.sun_path));
+
+	if (sun.sun_path[0] == '@')
+		sun.sun_path[0] = '\0';
```

Comment:
what's this do

Reply:
It's "standard" Linux syntax for _abstract unix sockets_.  ("standard" in quotation marks because several system tools use @ syntax for this but I can't find a reference for it anywhere.  Abstract sockets themselves are documented in `man 7 unix` with `\0` as the first character.)

I'll go add a breadcrumb comment so people reading the code have something to google.

Reply:
Ok, so systemd indicates the socket to use is an abstract socket by putting @ as the first character in the environment variable, we translate it into an abstract socket address by replacing with a byte. Fine then